### PR TITLE
Step7

### PR DESCRIPTION
**Pull request by ionicastefan20**

### CODE REVIEW

File: Step7/README.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
+# Step7
+Goal:
+* Download the Custom Vision model
+* Use it locally with tensorflow.js
+
+## Download the model locally
+
+- Press Export
```
Comment (pompomon) [inline Step7/README.md:12]:
```suggestion
- When creating Custom Vision project select a Compact pre-trained model (<SCREENSHOT>)

- Press Export

```

File: Step7/README.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+- Select Tensorflow.js and press Download
+
+![Select Tensorflow.js and press Download](../assets/screenshots/0_export_model_3.png "Select Tensorflow.js and press Download")
+
+- Extract all the files in `Step7/tf-js-model`
+
+- In the folder `Step7` run
+
+```
+npm install @microsoft/customvision-tfjs-node
```
Comment (pompomon) [inline Step7/README.md]: We can skip that and just add needed package to package.json right away, so people would get it on npm install

File: Step7/README.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+const fs = require('fs');
+const express = require('express');
+const app = express();
+const PORT = process.env.PORT || 1337;
+const bodyParser = require('body-parser')
+const cvstfjs = require("@microsoft/customvision-tfjs-node");
+
+...
+
+app.post('/predict', async (request, response) => {
```
Comment (pompomon) [inline Step7/README.md]: Let's have a separate endpoint for the offline prediction, like `/predict_offline`

File: Step7/README.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    const imageData = request.body;
+
+    let labels;
+    try {
+        const data = (await fs.readFileSync('tf-js-model/labels.txt')).toString();
```
Comment (pompomon) [inline Step7/README.md]: Let's init the model once on line 46, so we don't have to wait for the model to load weights each time we want to predict something

File: Step7/README.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        labels = data.split('\n');
+    } catch (err) {
+        console.log(err);
+        return;
```
Comment (pompomon) [inline Step7/README.md]: Let's make sure the app.js logic get the error state
```suggestion
    response.status(500);
    response.end();
    return;
```

File: Step7/README.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    }
+    
+    const model = new cvstfjs.ClassificationModel();
+    try {
+        await model.loadModelAsync('file://tf-js-model/model.json');
+    } catch (err) {
+        console.log(err)
+        return;
```
Comment (pompomon) [inline Step7/README.md:64]: Let's make sure the app.js logic get the error state
```suggestion
    response.status(500);
    response.end();
    return;
```

File: Step7/README.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        labels = data.split('\n');
+    } catch (err) {
+        console.log(err);
+        return;
+    }
+    
+    const model = new cvstfjs.ClassificationModel();
+    try {
+        await model.loadModelAsync('file://tf-js-model/model.json');
+    } catch (err) {
```
Comment (pompomon) [inline Step7/README.md]: Let's init the model once on line 46, so we don't have to wait for the model to load weights each time we want to predict something.
Bonus points - to have the endpoint fallback to the Azure Custom Vision service if the model failed to load

File: Step7/README.md [hunk 8]
```diff
@@ -0,0 +1,100 @@
+# Step7
+Goal:
+* Download the Custom Vision model
+* Use it locally with tensorflow.js
+
+## Download the model locally
+
+- When creating Custom Vision project select a Compact pre-trained model (<SCREENSHOT>)
```
Comment (pompomon) [inline Step7/README.md:8]: Please don't forget to add the screenshot ;)

File: Step7/index.js [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+(async () => {
+    let labels;
+    try {
+        const data = fs.readFileSync('tf-js-model/labels.txt').toString();
+        labels = data.split('\n');
+    } catch {
+        console.log("Could not load labels.txt");
+    }
+    if (!labels)
+        return;
```
Comment (pompomon) [inline Step7/index.js]:
```suggestion
```
image

File: Step7/README.md [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+(async () => {
+    let labels;
+    try {
+        const data = fs.readFileSync('tf-js-model/labels.txt').toString();
+        labels = data.split('\n');
+    } catch {
+        console.log("Could not load labels.txt");
+    }
+    if (!labels)
+        return;
```
Comment (pompomon) [inline Step7/README.md:58]:
```suggestion
    if (!labels || labels.length === 0) {
        console.log("No labels provided")
        return;
    }
```

**Review by pompomon — APPROVED** [review]
Looks good to me, please resolve the opened comments and I think it's good to be merged :)
